cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java

```diff
@@ -193,9 +205,44 @@ public void doAction(@Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices
         if (returnMap == null) {
             throw new RuntimeException("Action method did not return Map<String,String>");
         }
+
+        if (hasException(returnMap)) {
+            logException(returnMap);
+        }
+
         return handleSensitiveValues(returnMap, currentContext);
     }
 
+    private boolean hasException(Map<String, Serializable> returnMap) {
+        return returnMap.containsKey(RETURN_CODE) && !"0".equals(returnMap.get(RETURN_CODE)) &&
+            returnMap.containsKey(EXCEPTION);
+    }
+
+    private void logException(Map<String, Serializable> returnMap) {
+
+        String stacktrace = (String) returnMap.get(EXCEPTION);
+
+        if (REMOVE_MESSAGE_FROM_LOGGED_EX) {
+            StringWriter writer = new StringWriter();
+            final PrintWriter printWriter = new PrintWriter(writer);
+
+            BufferedReader reader = new BufferedReader(new StringReader(stacktrace));
+
+            reader.lines().forEach(line -> {
+                int idx = line.indexOf(MARKER);
+                String str = idx == -1 ? line : line.substring(0, idx + MARKER.length());
+
+                printWriter.println(str);
+            });
+
+            printWriter.close();
```

Comment:
should not close printWriter in memory as it does nothing.

Reply:
there may be unflushed data. It doesn't harm either.